File: bindings/go/fixtures_test.go
// Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.

package triepack

import (
	"bytes"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

func fixtureDir() string {
	_, filename, _, _ := runtime.Caller(0)
	return filepath.Join(filepath.Dir(filename), "..", "..", "tests", "fixtures")
}

func readFixture(t *testing.T, name string) []byte {
	t.Helper()
	path := filepath.Join(fixtureDir(), name)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read fixture %s: %v", name, err)
	}
	return data
}

// ── Decode C-generated fixtures ─────────────────────────────────────

func TestDecodeFixtureEmpty(t *testing.T) {
	buf := readFixture(t, "empty.trp")
	result, err := Decode(buf)
	if err != nil {
		t.Fatalf("Decode failed: %v", err)
	}
	if len(result) != 0 {
		t.Fatalf("expected empty map, got %d entries", len(result))
	}
}

func TestDecodeFixtureSingleNull(t *testing.T) {
	buf := readFixture(t, "single_null.trp")
	result, err := Decode(buf)
	if err != nil {
		t.Fatalf("Decode failed: %v", err)
	}
	if len(result) != 1 {
		t.Fatalf("expected 1 entry, got %d", len(result))
	}
	v, ok := result["hello"]
	if !ok {
		t.Fatal("missing key 'hello'")
	}
	if v != nil {
		t.Fatalf("expected nil, got %v", v)
	}
}

func TestDecodeFixtureSingleInt(t *testing.T) {
	buf := readFixture(t, "single_int.trp")
	result, err := Decode(buf)
	if err != nil {
		t.Fatalf("Decode failed: %v", err)
	}
	assertIntValue(t, result, "key", 42)
}

func TestDecodeFixtureMultiMixed(t *testing.T) {
	buf := readFixture(t, "multi_mixed.trp")
	result, err := Decode(buf)
	if err != nil {
		t.Fatalf("Decode failed: %v", err)
	}
	if result["bool"] != true {
		t.Fatalf("expected bool=true, got %v", result["bool"])
	}
	f, ok := result["f64"].(float64)
	if !ok || math.Abs(f-3.14159) > 1e-10 {
		t.Fatalf("expected f64~3.14159, got %v", result["f64"])
	}
	assertIntValue(t, result, "int", -100)
	if result["str"] != "hello" {
		t.Fatalf("expected str='hello', got %v", result["str"])
	}
	assertIntValue(t, result, "uint", 200)
}

func TestDecodeFixtureSharedPrefix(t *testing.T) {
	buf := readFixture(t, "shared_prefix.trp")
	result, err := Decode(buf)
	if err != nil {
		t.Fatalf("Decode failed: %v", err)
	}
	assertIntValue(t, result, "abc", 10)
	assertIntValue(t, result, "abd", 20)
	assertIntValue(t, result, "xyz", 30)
}

func TestDecodeFixtureLarge(t *testing.T) {
	buf := readFixture(t, "large.trp")
	result, err := Decode(buf)
	if err != nil {
		t.Fatalf("Decode failed: %v", err)
	}
	if len(result) != 100 {
		t.Fatalf("expected 100 entries, got %d", len(result))
	}
	for i := 0; i < 100; i++ {
		key := fmt.Sprintf("key_%04d", i)
		assertIntValue(t, result, key, i)
	}
}

func TestDecodeFixtureKeysOnly(t *testing.T) {
	buf := readFixture(t, "keys_only.trp")
	result, err := Decode(buf)
	if err != nil {
		t.Fatalf("Decode failed: %v", err)
	}
	for _, key := range []string{"apple", "banana", "cherry"} {
		v, ok := result[key]
		if !ok {
			t.Fatalf("missing key '%s'", key)
		}
		if v != nil {
			t.Fatalf("expected nil for '%s', got %v", key, v)
		}
	}
}

// ── Byte-for-byte binary reproducibility ────────────────────────────

func TestEncodeEmptyMatchesFixture(t *testing.T) {
	goBuf, err := Encode(map[string]interface{}{})
	if err != nil {
		t.Fatalf("Encode failed: %v", err)
	}
	cBuf := readFixture(t, "empty.trp")
	if !bytes.Equal(goBuf, cBuf) {
		t.Fatalf("Go output (%d bytes) != C fixture (%d bytes)", len(goBuf), len(cBuf))
	}
}

func TestEncodeSingleNullMatchesFixture(t *testing.T) {
	goBuf, err := Encode(map[string]interface{}{"hello": nil})
	if err != nil {
		t.Fatalf("Encode failed: %v", err)
	}
	cBuf := readFixture(t, "single_null.trp")
	if !bytes.Equal(goBuf, cBuf) {
		t.Fatalf("Go output (%d bytes) != C fixture (%d bytes)", len(goBuf), len(cBuf))
	}
}

func TestEncodeSingleIntMatchesFixture(t *testing.T) {
	goBuf, err := Encode(map[string]interface{}{"key": 42})
	if err != nil {
		t.Fatalf("Encode failed: %v", err)
	}
	cBuf := readFixture(t, "single_int.trp")
	if !bytes.Equal(goBuf, cBuf) {
		t.Fatalf("Go output (%d bytes) != C fixture (%d bytes)", len(goBuf), len(cBuf))
	}
}

func TestEncodeMultiMixedMatchesFixture(t *testing.T) {
	goBuf, err := Encode(map[string]interface{}{
		"bool": true,
		"f64":  3.14159,
		"int":  -100,
		"str":  "hello",
		"uint": 200,
	})
	if err != nil {
		t.Fatalf("Encode failed: %v", err)
	}
	cBuf := readFixture(t, "multi_mixed.trp")
	if !bytes.Equal(goBuf, cBuf) {
		t.Fatalf("Go output (%d bytes) != C fixture (%d bytes)", len(goBuf), len(cBuf))
	}
}

func TestEncodeSharedPrefixMatchesFixture(t *testing.T) {
	goBuf, err := Encode(map[string]interface{}{"abc": 10, "abd": 20, "xyz": 30})
	if err != nil {
		t.Fatalf("Encode failed: %v", err)
	}
	cBuf := readFixture(t, "shared_prefix.trp")
	if !bytes.Equal(goBuf, cBuf) {
		t.Fatalf("Go output (%d bytes) != C fixture (%d bytes)", len(goBuf), len(cBuf))
	}
}

func TestEncodeLargeMatchesFixture(t *testing.T) {
	data := make(map[string]interface{})
	for i := 0; i < 100; i++ {
		data[fmt.Sprintf("key_%04d", i)] = i
	}
	goBuf, err := Encode(data)
	if err != nil {
		t.Fatalf("Encode failed: %v", err)
	}
	cBuf := readFixture(t, "large.trp")
	if !bytes.Equal(goBuf, cBuf) {
		t.Fatalf("Go output (%d bytes) != C fixture (%d bytes)", len(goBuf), len(cBuf))
	}
}

func TestEncodeKeysOnlyMatchesFixture(t *testing.T) {
	goBuf, err := Encode(map[string]interface{}{"apple": nil, "banana": nil, "cherry": nil})
	if err != nil {
		t.Fatalf("Encode failed: %v", err)
	}
	cBuf := readFixture(t, "keys_only.trp")
	if !bytes.Equal(goBuf, cBuf) {
		t.Fatalf("Go output (%d bytes) != C fixture (%d bytes)", len(goBuf), len(cBuf))
	}
}

File: bindings/typescript/tests/fixtures.test.ts
// Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.

import * as fs from "fs";
import * as path from "path";
import { encode, decode, TriePackData } from "../src/index";

const FIXTURE_DIR = path.resolve(__dirname, "../../../tests/fixtures");

function readFixture(name: string): Uint8Array {
    return new Uint8Array(fs.readFileSync(path.join(FIXTURE_DIR, name)));
}

describe("cross-language fixture tests", () => {
    test("fixtures directory exists", () => {
        expect(fs.existsSync(FIXTURE_DIR)).toBe(true);
    });

    describe("decode C-generated fixtures", () => {
        test("empty.trp -> {} (0 keys)", () => {
            const result = decode(readFixture("empty.trp"));
            expect(result).toEqual({});
        });

        test("single_null.trp -> { hello: null }", () => {
            const result = decode(readFixture("single_null.trp"));
            expect(result).toEqual({ hello: null });
        });

        test("single_int.trp -> { key: 42 }", () => {
            const result = decode(readFixture("single_int.trp"));
            expect(result).toEqual({ key: 42 });
        });

        test("multi_mixed.trp -> 5 mixed-type keys", () => {
            const result = decode(readFixture("multi_mixed.trp"));
            expect(result.bool).toBe(true);
            expect(result.f64 as number).toBeCloseTo(3.14159, 10);
            expect(result.int).toBe(-100);
            expect(result.str).toBe("hello");
            expect(result.uint).toBe(200);
        });

        test("shared_prefix.trp -> { abc: 10, abd: 20, xyz: 30 }", () => {
            const result = decode(readFixture("shared_prefix.trp"));
            expect(result).toEqual({ abc: 10, abd: 20, xyz: 30 });
        });

        test("large.trp -> 100 sequential keys", () => {
            const result = decode(readFixture("large.trp"));
            expect(Object.keys(result).length).toBe(100);
            for (let i = 0; i < 100; i++) {
                const key = "key_" + String(i).padStart(4, "0");
                expect(result[key]).toBe(i);
            }
        });

        test("keys_only.trp -> { apple: null, banana: null, cherry: null }", () => {
            const result = decode(readFixture("keys_only.trp"));
            expect(result).toEqual({ apple: null, banana: null, cherry: null });
        });
    });

    describe("byte-for-byte binary reproducibility", () => {
        test("empty -> matches C fixture", () => {
            const tsBuf = encode({});
            const cBuf = readFixture("empty.trp");
            expect(Buffer.from(tsBuf).equals(Buffer.from(cBuf))).toBe(true);
        });

        test("single_null -> matches C fixture", () => {
            const tsBuf = encode({ hello: null });
            const cBuf = readFixture("single_null.trp");
            expect(Buffer.from(tsBuf).equals(Buffer.from(cBuf))).toBe(true);
        });

        test("single_int -> matches C fixture", () => {
            const tsBuf = encode({ key: 42 });
            const cBuf = readFixture("single_int.trp");
            expect(Buffer.from(tsBuf).equals(Buffer.from(cBuf))).toBe(true);
        });

        test("multi_mixed -> matches C fixture", () => {
            const data: TriePackData = {
                bool: true,
                f64: 3.14159,
                int: -100,
                str: "hello",
                uint: 200,
            };
            const tsBuf = encode(data);
            const cBuf = readFixture("multi_mixed.trp");
            expect(Buffer.from(tsBuf).equals(Buffer.from(cBuf))).toBe(true);
        });

        test("shared_prefix -> matches C fixture", () => {
            const tsBuf = encode({ abc: 10, abd: 20, xyz: 30 });
            const cBuf = readFixture("shared_prefix.trp");
            expect(Buffer.from(tsBuf).equals(Buffer.from(cBuf))).toBe(true);
        });

        test("large -> matches C fixture", () => {
            const data: TriePackData = {};
            for (let i = 0; i < 100; i++) {
                data["key_" + String(i).padStart(4, "0")] = i;
            }
            const tsBuf = encode(data);
            const cBuf = readFixture("large.trp");
            expect(Buffer.from(tsBuf).equals(Buffer.from(cBuf))).toBe(true);
        });

        test("keys_only -> matches C fixture", () => {
            const tsBuf = encode({ apple: null, banana: null, cherry: null });
            const cBuf = readFixture("keys_only.trp");
            expect(Buffer.from(tsBuf).equals(Buffer.from(cBuf))).toBe(true);
        });
    });
});
